src/main/java/seedu/address/model/Model.java

```diff
@@ -91,4 +94,47 @@
      * @throws NullPointerException if {@code predicate} is null.
      */
     void updateFilteredModuleList(Predicate<Module> predicate);
+
+    // ============================ TodoList ==================================================
+
+    /**
+     * Replaces todo list data with the data in {@code todolist}.
+     */
+    void setTodoList(ReadOnlyTodoList todoList);
+
+    /** Returns the TodoList */
+    ReadOnlyTodoList getTodoList();
+
+    /**
+     * Returns true if a task with the same name, date, and type as {@code task} exists in the todo list.
+     */
+    boolean hasTask(Task task);
+
+    /**
+     * Deletes the given task.
+     * The task must exist in the todo list.
+     */
+    void deleteTask(Task target);
+
+    /**
+     * Adds the given task.
+     * {@code task} must not already exist in the todo list.
+     */
+    void addTask(Task task);
+
+    /**
+     * Replaces the given task {@code target} with {@code editedTask}.
+     * {@code target} must exist in the todo list.
+     * The task name of {@code editedTask} must not be the same as another existing task in the todo list.
+     */
+    void setTask(Task target, Task editedTask);
```

Comment:
Should we ensure that another task with the same name, date and type as the edited task does not exist in the todo list? This is in accordance with the hasTask method above which checks if there is a task with the same name, date and type.

Reply:
Yes, I think we should handle that in each of the Command class?

In addition, I just changed the isSameTask() method. Now 2 tasks are considered to be the same if only their name are the same since the other fields can be optional. I think it is sufficient to just check the name?

Reply:
Yes, I think it is sufficient to check the name for duplicate tasks. However, could you clarify the implementation of the above hasTask method? The javadocs says that it will check if there is an existing task with the same name, date, and type in the todo list. In that case shouldn't checking for same name be enough?

Reply:
Oh yes, it should only need to have the same name. I forgot to edit the JavaDoc for that one 😆. Thanks for pointing this out!

Reply:
Okay, I have updated the JavaDocs.